[Refactor] Extracted cage into a standalone library

Yaxuan-w opened the pull request:
This PR extracts `Cage` struct and corresponding `vmmap` + `signal` implementation from obsolete RawPOSIX. This refinement would help avoiding circular imports in upcoming 3i implementations, since both 3i and rawposix need to access cage struct, and 3i will send calls to rawposix.

rennergade reviewed (commented):
This looks great to me and is really clean.

My only concern is being very diligent to make sure it's updated to the current changes in terms of memory stuff when we're ready to merge. For example we just added shared memory functions.

Additionally do we want to add shm.rs to this library? It might make sense conceptually.

Yaxuan-w commented:
> My only concern is being very diligent to make sure it's updated to the current changes in terms of memory stuff when we're ready to merge. For example we just added shared memory functions.

Thanks for reminding! I added this as a todo on Issue #135. We also need to integrate signal stuffs (which @celinehoang177 is working on). 

> Additionally do we want to add shm.rs to this library? It might make sense conceptually.

Yes

lukpueh commented:
Is it worth adding some details, where this was extracted from? Is there any previous git history worth preserving?

rennergade commented:
My previous comments still stand here but this is also good from my side.

JustinCappos reviewed (changes requested):
Overall, this is pretty good.  A few high level comments (in addition to my notes on the PR).  

It would often be useful to explain why a function / block of code / constant is useful in addition to what it does.  Just as a general programming tip, that extra few words can help an outside reader really understand what you're doing.  It also helps a reviewer understand if you were or were not anticipating a use case of this function.

Are there tests for this?  I'd love to see unit tests for a lot of these functions.   Given this is also set up as an interface so different implementations can be used, performance benchmarking (likely through criterion, as I did in fdtables), would be useful too.

For benchmarking, it would be fine to merge this without and to open an issue to add those after.  I'm hoping we can get at least some rudimentary tests in place before merging though.

rennergade commented on src/cage/src/cage.rs:
```diff
@@ -0,0 +1,121 @@
+    // Current working directory of cage, must be able to be unique from other cages
+    pub cwd: RwLock<Arc<PathBuf>>,
+    // Identifiers for gid/uid/egid/euid
+    pub gid: AtomicI32,
+    pub uid: AtomicI32,
+    pub egid: AtomicI32,
+    pub euid: AtomicI32,
```
https://github.com/Lind-Project/lind-wasm/issues/327

rennergade commented on src/cage/src/memory/vmmap.rs:
```diff
@@ -0,0 +1,950 @@
+        self.entries.first_key_value()
+    }
+
+    /// Creates a double-ended iterator over all entries
```
as far as I can tell this was added as extra functionality thats not used anywhere. Remove it?

rennergade commented on src/cage/src/memory/mem_helper.rs:
```diff
@@ -0,0 +1,206 @@
+    child_vmmap.set_program_break(parent_vmmap.program_break);
+}
+
+/// Validates and converts a virtual memory address to a physical address with protection checks
```
this also seems to be unused but I'm going to check if this changes in dennis's PR

Yaxuan-w commented:
Why we're removing files from RawPOSIX in this PR?

Yaxuan-w commented on src/cage/src/cage.rs:
```diff
@@ -0,0 +1,141 @@
+///
+/// Pre-allocate MAX_CAGEID elements, all initialized to None.
+/// Lazy causes `CAGE_MAP` to be initialized when it is first accessed, rather than when the program starts.
+// pub static CAGE_MAP: Lazy<RwLock<Vec<Option<Arc<Cage>>>>> = Lazy::new(|| {
+//     let mut vec = Vec::with_capacity(MAX_CAGEID);
+//     vec.resize_with(MAX_CAGEID, || None);
+//     RwLock::new(vec)
+// });
```
Remove comments

Yaxuan-w commented on src/cage/src/memory/shared.rs:
```diff
@@ -0,0 +1,238 @@
+
+use libc::*;
+
+pub static SHM_METADATA: interface::RustLazyGlobal<interface::RustRfc<ShmMetadata>> =
```
Replace `interface::RustLazyGlobal` and `interface::RustRfc`

Yaxuan-w commented on src/cage/src/memory/shared.rs:
```diff
@@ -0,0 +1,238 @@
+    pub shminfo: ShmidsStruct,
+    pub key: i32,
+    pub size: usize,
+    pub filebacking: interface::ShmFile,
```
Replace `interface::`

Yaxuan-w commented on src/cage/src/memory/shared.rs:
```diff
@@ -0,0 +1,238 @@
+    pub size: usize,
+    pub filebacking: interface::ShmFile,
+    pub rmid: bool,
+    pub attached_cages: interface::RustHashMap<u64, i32>, // attached cages, number of references in cage
```
Replace `interface::`

Yaxuan-w commented on src/cage/src/memory/shared.rs:
```diff
@@ -0,0 +1,238 @@
+            size: size,
+            filebacking: filebacking,
+            rmid: false,
+            attached_cages: interface::RustHashMap::new(),
```
Replace `interface::`

Yaxuan-w commented on src/cage/src/memory/shared.rs:
```diff
@@ -0,0 +1,238 @@
+        self.shminfo.shm_atime = timestamp() as isize;
+
+        match self.attached_cages.entry(cageid) {
+            interface::RustHashEntry::Occupied(mut occupied) => {
```
Replace `interface::`

Yaxuan-w commented on src/cage/src/memory/shared.rs:
```diff
@@ -0,0 +1,238 @@
+            interface::RustHashEntry::Occupied(mut occupied) => {
+                *occupied.get_mut() += 1;
+            }
+            interface::RustHashEntry::Vacant(vacant) => {
```
Replace `interface::`

Yaxuan-w commented on src/cage/src/memory/shared.rs:
```diff
@@ -0,0 +1,238 @@
+                vacant.insert(1);
+            }
+        };
+        interface::libc_mmap(
```
Replace `interface::`. If we call `mmap` here, it will cause circular imports

Yaxuan-w commented on src/cage/src/memory/shared.rs:
```diff
@@ -0,0 +1,238 @@
+        self.shminfo.shm_nattch -= 1;
+        self.shminfo.shm_dtime = timestamp() as isize;
+        match self.attached_cages.entry(cageid) {
+            interface::RustHashEntry::Occupied(mut occupied) => {
```
Replace `interface::`

Yaxuan-w commented on src/cage/src/memory/shared.rs:
```diff
@@ -0,0 +1,238 @@
+                    occupied.remove_entry();
+                }
+            }
+            interface::RustHashEntry::Vacant(_) => {
```
Replace `interface::`

Yaxuan-w commented on src/cage/src/memory/memory.rs:
```diff
@@ -0,0 +1,202 @@
+/// # Returns
+/// * `Ok(u64)` - Translated physical memory address
+
+pub fn translate_vmmap_addr(cage: &Cage, arg: u64) -> Result<u64, Errno> {
```
This should be in type conversion lib

Yaxuan-w commented on src/cage/src/memory/memory.rs:
```diff
@@ -0,0 +1,202 @@
+/// - Ensuring addresses are properly aligned to pages
+/// - Validating all pages in the region are mapped with correct permissions
+/// - Preventing access outside of allocated memory regions
+pub fn check_and_convert_addr_ext(
```
This should be in type conversion lib

Yaxuan-w commented on src/cage/src/memory/memory.rs:
```diff
@@ -0,0 +1,202 @@
+                );
+
+                // revert child's memory region prot
+                libc::mprotect(child_st as *mut libc::c_void, addr_len, entry.prot);
```
this will cause circular import

Yaxuan-w commented on src/cage/src/memory/memory.rs:
If we want to have a separate helper function file for vmmap, it would make more sense to put inside rawposix, or separate into another standalone vmmap lib to avoid circular import

Yaxuan-w commented on src/cage/src/memory/mod.rs:
```diff
@@ -0,0 +1,9 @@
+
+pub use memory::*;
+pub use vmmap::*;
+pub use shared::*;
```
add new line

Yaxuan-w commented:
This looks good to me now.
